### DISCUSSION TITLE
Issue with a custom function

### DISCUSSION THREAD

**Discussion by divya-akula** [microsoft/hack-together-teams#68] [Q&A]
Hi ,

I am facing the following issue, stating _The DefaultPromptManager.invokeFunction() method was called for an unregistered function named "GetUsers"_ though I have registered it in the second screenshot, I am stuck with this , any help wold be highly appreciated.

![image](https://github.com/microsoft/hack-together-teams/assets/9211327/e879c6f2-c3a3-46ff-9ec9-bf8ef6f0a33f)

though I have registered the action
![image](https://github.com/microsoft/hack-together-teams/assets/9211327/7e0dee73-5b0b-4be1-b734-309556b08a20)

**Comment by singhk97** [comment]
You should be doing this instead:

```javascript
// Add a prompt function for getting the current status of the lights
app.ai.prompts.addFunction(
    'GetUsers',
    async (context: TurnContext, state: ApplicationTurnState) => {
        getUsers(state, ...);
        return true;
    }
);
```

Note that you cannot pass `UserData` as a parameter, the only allowed parameters are `TurnContext` and `ApplicationTurnState`. There are a few ways you can do that.

If it's static, independent of the turn & read only, define it as a variable in the scope in which you're defining the prompt function and use it in the handler function. Javascript's closure should work. Note that this approach can cause concurrency issues.

Another, more appropriate way to access `UserData` is to inject it into the turn state `ApplicationTurnState` before the turn like this:

```javascript
app.turn('beforeTurn', (context, state) => {
    state.temp.userData.value = // assign it the user data
    // NOTE: you will have to update the typescript interface appropriately
});
```

```javascript
// inside the handler function for the `GetUsers` prompt function registration access the data like this
state.temp.userData.value
```
